platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/managed/ManagedObjectRegistry.java

```diff
@@ -0,0 +1,220 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.instantiation.managed;
+
+import org.gradle.api.model.ManagedObjectCreator;
+import org.gradle.api.model.ManagedObjectProvider;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.service.AnnotatedServiceLifecycleHandler;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+import org.jspecify.annotations.Nullable;
+
+import java.lang.annotation.Annotation;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Responsible for discovering services that provide Gradle-managed types, and creating instances of those types.
+ * <p>
+ * Discovers all services annotated with {@link ManagedObjectProvider} and searches its methods for those annotated
+ * with {@link ManagedObjectCreator}.
+ */
+@ServiceScope({Scope.Global.class, Scope.BuildTree.class, Scope.Build.class, Scope.Project.class})
+public class ManagedObjectRegistry implements AnnotatedServiceLifecycleHandler {
+
+    private final ManagedObjectRegistry parent;
+    private final Map<Class<?>, FactoryRegistration> factories = new ConcurrentHashMap<>();
+
+    public ManagedObjectRegistry(@Nullable ManagedObjectRegistry parent) {
+        this.parent = parent;
+    }
+
+    @Override
+    public List<Class<? extends Annotation>> getAnnotations() {
+        return Collections.singletonList(ManagedObjectProvider.class);
+    }
+
+    @Nullable
+    @Override
+    public Class<? extends Annotation> getImplicitAnnotation() {
+        return null;
+    }
+
+    @Override
+    public void whenRegistered(Class<? extends Annotation> annotation, Registration registration) {
+        assert annotation == ManagedObjectProvider.class;
+        Object instance = registration.getInstance();
+
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+        for (Method method : instance.getClass().getMethods()) {
+            if (method.isAnnotationPresent(ManagedObjectCreator.class) ||
+                doesParentInterfaceMethodHaveAnnotation(instance.getClass(), method, ManagedObjectCreator.class)
+            ) {
+                if (method.getReturnType() == Void.class) {
+                    throw new IllegalArgumentException("Method " + method + " annotated with @ManagedObjectCreator must return a value");
+                }
+
+                MethodHandle handle;
+                try {
+                     handle = lookup.unreflect(method);
+                } catch (IllegalAccessException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                }
+
+                registerFactory(handle, instance);
```

Comment:
❔   Should we fail if the provider classes did not register any creator-methods to prevent those from lingering around?

Reply:
Good idea. Added a check